### PR TITLE
mariadb: update packages

### PR DESCRIPTION
**Pull request by Izorkin**
###### Motivation for this change
Move packages connector-c and galera to external. Need to will the package appear on https://nixos.org/nixos/packages.html
Update package galera to v25.3.27
Add packages:
- connector-c v3.1.2
- galera v26.4.2

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by FRidh** [comment]
In order to achieve that one could also simply use `inherit (foo) bar;`. Was this approach agreed upon somewhere?

**Comment by Izorkin** [comment]
@FRidh nowhere approach agreed.
After wanted to divide mariadb into separate packages to mariadb-server and mariadb-client
https://github.com/NixOS/nixpkgs/issues/64132
https://github.com/NixOS/nixpkgs/issues/63723

**Comment by Izorkin** [comment]
> In order to achieve that one could also simply use `inherit (foo) bar;`
I don't know how to use this variant

**Comment by FRidh** [comment]
@grahamcofborg eval

**Comment by Izorkin** [comment]
@FRidh not need rename mysql.connector-c to libmysql? Probably it will be so clearer.

**Comment by Izorkin** [comment]
@GrahamcOfBorg eval

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb-galera

**Comment by Mic92** [comment]
@GrahamcOfBorg build mariadb-galera

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb-galera_26 mariadb-connector-c_3_0 mariadb-connector-c_3_1

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb-galera_26 mariadb-connector-c_3_0 mariadb-connector-c_3_1

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb-connector-c_2_3 mariadb-connector-c_3_0 mariadb-connector-c_3_1

**Comment by Mic92** [comment]
@GrahamcOfBorg build mariadb-connector-c_2_3 mariadb-connector-c_3_0 mariadb-connector-c_3_1

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -14653,8 +14653,18 @@ in
 
   rpcbind = callPackage ../servers/rpcbind { };
 
-  mariadb = callPackage ../servers/sql/mariadb {
+  libmysql = mariadb-connector-c;
+  mariadb-connector-c = mariadb-connector-c_2_3;
+  mariadb-connector-c_2_3 = callPackage ../servers/sql/mariadb/connector-c/2_3.nix { };
+  mariadb-connector-c_3_0 = callPackage ../servers/sql/mariadb/connector-c/3_0.nix { };
+  mariadb-connector-c_3_1 = callPackage ../servers/sql/mariadb/connector-c/3_1.nix { };
```
Comment (Mic92) [inline pkgs/top-level/all-packages.nix]: Any reason to have 3.0, when we already have 3.1?

Comment (Mic92) [reply pkgs/top-level/all-packages.nix]: Could we also just drop 2.3? Archlinux does this as well.

Comment (Izorkin) [reply pkgs/top-level/all-packages.nix]: I think first leave this variant. In the next stage change to 3.0 / 3.1

Comment (Mic92) [reply pkgs/top-level/all-packages.nix]: For which reasons?

Comment (Izorkin) [reply pkgs/top-level/all-packages.nix]: Then it will be easier to check package builds via nix-review

Comment (Mic92) [reply pkgs/top-level/all-packages.nix]: You can only build a subset in `nix-review` by using the `-p` option.

Comment (Izorkin) [reply pkgs/top-level/all-packages.nix]: with
`nix-review pr -p mariadb-connector-c -p libmysql 65222`
checked only mariadb-connector-c libmysql

Comment (Mic92) [reply pkgs/top-level/all-packages.nix]: @Izorkin what would you expect instead?

Comment (Izorkin) [reply pkgs/top-level/all-packages.nix]: @Mic92 automatic build of packages that depend on libmysql

**Comment by Izorkin** [comment]
@GrahamcOfBorg build grass kexi libreoffice glsurf kodi gerbil neko urweb haskellPackages.mysql php72 php73 cppdb gdal gdal_2 libagar libdbiDriversBase libgda redland opendbx poco tntdb unixODBCDrivers.mariadb wt3 wt4 lispPackages.clsql lua51Packages.luadbi-mysql lua52Packages.luadbi-mysql lua53Packages.luadbi-mysql luajitPackages.luadbi-mysql ocamlPackages.ocaml_mysql perlPackages.DBDmysql purePackages.glpk pythonPackages.MySQL_python pythonPackages.mysqlclient rPackages.RMySQL sysbench tora clickhouse slurm freeradius lighttpd dovecot opensmtpd postfix zabbix uwsgi mydumper kea mailutils thc-hydra collectd

File: pkgs/development/libraries/libdbi-drivers/default.nix [hunk 2]
```diff
@@ -28,8 +28,8 @@ stdenv.mkDerivation rec {
     "--with-dbi-libdir=${libdbi}/lib"
   ] ++ optionals (mysql != null) [
```
Comment (aanderse) [inline pkgs/development/libraries/libdbi-drivers/default.nix]: -> `(libmysql != null)`

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -3758,7 +3758,7 @@ in {
 
   monotonic = callPackage ../development/python-modules/monotonic { };
 
-  MySQL_python = callPackage ../development/python-modules/mysql_python { };
+  MySQL_python = callPackage ../development/python-modules/mysql_python { libmysql = pkgs.mariadb-connector-c_2_3; };
```
Comment (aanderse) [inline pkgs/top-level/python-packages.nix]: Why are you inconsistently referring to the package? Shouldn't it be `libmysql` everywhere?

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -13656,7 +13656,7 @@ in
 
   unixODBC = callPackage ../development/libraries/unixODBC { };
 
-  unixODBCDrivers = recurseIntoAttrs (callPackages ../development/libraries/unixODBCDrivers {});
+  unixODBCDrivers = recurseIntoAttrs (callPackages ../development/libraries/unixODBCDrivers { libmysql = mariadb-connector-c_2_3; });
```
Comment (aanderse) [inline pkgs/top-level/all-packages.nix]: Why are you inconsistently referring to the package? Shouldn't it be `libmysql` everywhere?

**Review by aanderse — CHANGES_REQUESTED** [review]
If we're renaming the package to `libmysql` I think the directory structure should reflect that. I might even suggest putting a `libmysql` folder alongside `mariadb`.

Speaking of which, why are we renaming it to `libmysql`? Upstream, Debian, and Arch all call it `libmysqlclient`.

We're getting closer to `19.09` and this change will almost certainly cause some minor breakage. Do we feel like we have enough breathing room to deal with any problems that might come up from this PR? Given past history I would say we have just under 2 months... so I think that is enough time.

File: pkgs/development/libraries/libdbi-drivers/default.nix [hunk 5]
```diff
@@ -1,5 +1,5 @@
 { stdenv, fetchurl, libdbi
-, mysql ? null
+, mysql ? null, libmysql
```
[the remark this answers is not in the document]
Comment (Izorkin) [reply pkgs/development/libraries/libdbi-drivers/default.nix]: with this variant error build libdbiDriversBase
`error: anonymous function at pkgs/development/libraries/libdbi-drivers/default.nix:1:1 called with unexpected argument 'mysql', at lib/customisation.nix:69:12`

File: pkgs/top-level/python-packages.nix (hunk 3 above)
Comment (Izorkin) [reply pkgs/top-level/python-packages.nix]: With mariadb connector-c v3.1 error build
```
builder for '/nix/store/8sfw1ql6ia1wznyshylk6j69d2wyq2h7-python2.7-MySQL-python-1.2.5.drv' failed with exit code 1; last 10 log lines:
  copying MySQLdb/constants/CLIENT.py -> build/lib.linux-x86_64-2.7/MySQLdb/constants
  running build_ext
  building '_mysql' extension
  creating build/temp.linux-x86_64-2.7
  gcc -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -Dversion_info=(1,2,5,'final',1) -D__version__=1.2.5 -I/nix/store/h258lj4fc791l0lzdz04bf9za715fvxz-mariadb-connector-c-3.1.2/include/mariadb -I/nix/store/h258lj4fc791l0lzdz04bf9za715fvxz-mariadb-connector-c-3.1.2/include/mariadb/mysql -I/nix/store/whb8vd9pmssdmzhqffzkc3vz6ih3q1yk-python-2.7.16/include/python2.7 -c _mysql.c -o build/temp.linux-x86_64-2.7/_mysql.o
  _mysql.c:44:10: fatal error: my_config.h: No such file or directory
   #include "my_config.h"
            ^~~~~~~~~~~~~
  compilation terminated.
  error: command 'gcc' failed with exit status 1
[0 built (1 failed), 0.0 MiB DL]
error: build of '/nix/store/8sfw1ql6ia1wznyshylk6j69d2wyq2h7-python2.7-MySQL-python-1.2.5.drv' failed
```

**Comment by Izorkin** [comment]
> If we're renaming the package to `libmysql` I think the directory structure should reflect that. I might even suggest putting a `libmysql` folder alongside `mariadb`.
> 
> Speaking of which, why are we renaming it to `libmysql`? Upstream, Debian, and Arch all call it `libmysqlclient`.

libmysql was also used before
https://github.com/NixOS/nixpkgs/blob/release-16.03/pkgs/top-level/all-packages.nix#L9787

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (Izorkin) [reply pkgs/top-level/all-packages.nix]: unixODBCDrivers.mariadb updated to 3.1.2

File: pkgs/development/libraries/libdbi-drivers/default.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (aanderse) [reply pkgs/development/libraries/libdbi-drivers/default.nix]: See `all-packages.nix`:
```
  libdbiDriversBase = libdbiDrivers.override {
    mysql = null;
    sqlite = null;
  };
```

The reference needs to be updated there as well: `mysql = null;` -> `libmysql = null;`.

File: pkgs/top-level/python-packages.nix (hunk 3 above)
Comment (aanderse) [reply pkgs/top-level/python-packages.nix]: I meant why aren't we calling it `libmysql_2_3` (or more accurately `libmysqlclient_2_3`)?

**Comment by aanderse** [comment]
> libmysql was also used before

And that naming was dropped, correct? Possibly because it was confusing. If the library is called `libmysqlclient` upstream, and Debian and Arch call it `libmysqlclient` I would suggest **if** we are renaming it we should follow suit.

What was your motivation to rename the package? Overall I do agree with the rename (or at least an alias) because everywhere I've ever seen the library is referred to as `libmysqlclient`.

**Comment by Izorkin** [comment]
@aanderse thanks, updated PR. Replaced to libmysqlclient

**Review by aanderse — CHANGES_REQUESTED** [review]
Minor nitpick.

**Comment by aanderse** [comment]
@Izorkin thanks for your hard work. It would be nice to have someone like @globin (marked as maintainer) to approve this as well.

We're still about ~2 months away from `19.09` so any breakage that could potentially occur from this should have plenty of time to be discovered and fixed.

What do you think @lheckemann and @disassembler?

**Comment by Izorkin** [comment]
@aanderse  thanks, fixed.

**Comment by lheckemann** [comment]
All I can think of is running some nixos tests, being not much of an expert on mysql.

@grahamcofborg test gitea

**Review by aanderse — APPROVED** [review]

**Comment by Izorkin** [comment]
Redo rebase again?

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb-connector-c_2_3 mariadb-connector-c_3_1 mariadb-galera_25
@GrahamcOfBorg test gitea mysql mysql-backup mysql-replication

**Comment by aanderse** [comment]
@Izorkin well @lheckemann didn't say no... so how about you sort out the remaining issues and then we merge it? We still have a few weeks until the freeze, so if we do this soon it should give us enough time to deal with any potential issues which could arise.

**Comment by Izorkin** [comment]
@aanderse i have not encountered other problems yet

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb-connector-c_2_3 mariadb-connector-c_3_1 mariadb-galera_25

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb-connector-c, mariadb-galera, unixODBCDrivers.mariadb

**Comment by Izorkin** [comment]
@GrahamcOfBorg test gitea mysql mysql-backup mysql-replication

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb-connector-c_2_3 mariadb-connector-c_3_1 mariadb-galera_25 unixODBCDrivers.mariadb
@GrahamcOfBorg test gitea mysql mysql-backup mysql-replication

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb-connector-c_3_1

**Comment by aanderse** [comment]
@Izorkin it looks like some tests aren't properly evaluating, if I'm not mistaken:

> The following builds were skipped because they don't evaluate on x86_64-linux: tests.mysql-backup, tests.mysql-replication

Any idea?

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mysql-backup mysql-replication

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mysqlBackup mysqlReplication
@GrahamcOfBorg test gitea mysql

**Comment by Izorkin** [comment]
Оn my server tests gitea mysql mysqlBackup correct work.
For the test mysqlReplication need this PR #61671

File: pkgs/servers/sql/mariadb/connector-c/2_3.nix [hunk 6]
```diff
@@ -0,0 +1,6 @@
+{ callPackage, ... } @ args:
+
+callPackage ./default.nix (args // {
```
Comment (disassembler) [inline pkgs/servers/sql/mariadb/connector-c/2_3.nix]: `./.` is the standard way to call `default.nix` in current directory.

File: pkgs/servers/sql/mariadb/connector-c/3_1.nix [hunk 7]
```diff
@@ -0,0 +1,6 @@
+{ callPackage, ... } @ args:
+
+callPackage ./default.nix (args // {
```
Comment (disassembler) [inline pkgs/servers/sql/mariadb/connector-c/3_1.nix]: `./.` is the standard way to call `default.nix` in current directory.

File: pkgs/servers/sql/mariadb/galera/25.nix [hunk 8]
```diff
@@ -0,0 +1,6 @@
+{ callPackage, ... } @ args:
+
+callPackage ./default.nix (args // {
```
Comment (disassembler) [inline pkgs/servers/sql/mariadb/galera/25.nix]: `./.` is the standard way to call `default.nix` in current directory.

File: pkgs/servers/sql/mariadb/galera/default.nix [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+, version, sha256, ...
+}:
+
+let
+  galeraLibs = buildEnv {
+    name = "galera-lib-inputs-united";
+    paths = [ openssl.out boost check ];
+  };
+
+in stdenv.mkDerivation rec {
```
Comment (disassembler) [inline pkgs/servers/sql/mariadb/galera/default.nix]: move pname into the let block and inherit it and then you can drop the rec on the attrSet.

File: pkgs/top-level/aliases.nix [hunk 10]
```diff
@@ -187,7 +187,7 @@ mapAliases ({
   libintlOrEmpty = stdenv.lib.optional (!stdenv.isLinux || stdenv.hostPlatform.libc != "glibc") gettext; # added 2018-03-14
   libjson_rpc_cpp = libjson-rpc-cpp; # added 2017-02-28
   liblapackWithoutAtlas = liblapack; # added 2018-11-05
-  libmysql = mysql.connector-c; # added # 2017-12-28, this was a misnomer refering to libmysqlclient
+  libmysql = mariadb-connector-c; # added # 2017-12-28, this was a misnomer refering to libmysqlclient
```
Comment (disassembler) [inline pkgs/top-level/aliases.nix]: is this comment still needed/accurate?

**Review by disassembler — CHANGES_REQUESTED** [review]

File: pkgs/servers/sql/mariadb/connector-c/2_3.nix (hunk 6 above)
Comment (Izorkin) [reply pkgs/servers/sql/mariadb/connector-c/2_3.nix]: Thanks, fixed

File: pkgs/top-level/aliases.nix (hunk 10 above)
Comment (Izorkin) [reply pkgs/top-level/aliases.nix]: not sure

File: pkgs/servers/sql/mariadb/galera/default.nix (hunk 9 above)
Comment (Izorkin) [reply pkgs/servers/sql/mariadb/galera/default.nix]: @disassembler that?
```
diff --git a/pkgs/servers/sql/mariadb/galera/default.nix b/pkgs/servers/sql/mariadb/galera/default.nix
index 3fb6ebfbfc9..263a21983e2 100644
--- a/pkgs/servers/sql/mariadb/galera/default.nix
+++ b/pkgs/servers/sql/mariadb/galera/default.nix
@@ -4,13 +4,14 @@
 }:

 let
+  pname = "mariadb-galera";
   galeraLibs = buildEnv {
     name = "galera-lib-inputs-united";
     paths = [ openssl.out boost check ];
   };

-in stdenv.mkDerivation rec {
-  pname = "mariadb-galera";
+in stdenv.mkDerivation {
+  inherit pname;
   inherit version;

   src = fetchFromGitHub {
```

File: pkgs/top-level/aliases.nix (hunk 10 above)
Comment (Izorkin) [reply pkgs/top-level/aliases.nix]: removed alias

**Comment by Izorkin** [comment]
Rebased PR

File: pkgs/top-level/python-packages.nix [hunk 11]
```diff
@@ -3873,7 +3873,9 @@ in {
 
   monotonic = callPackage ../development/python-modules/monotonic { };
 
-  MySQL_python = callPackage ../development/python-modules/mysql_python { };
+  MySQL_python = callPackage ../development/python-modules/mysql_python {
```
Comment (Mic92) [inline pkgs/top-level/python-packages.nix]: I opened a pr to remove this package: https://github.com/NixOS/nixpkgs/pull/68354

**Comment by Izorkin** [comment]
Thanks!

**Pull request merged**